src/Microsoft.Extensions.ObjectPool/DefaultObjectPool.cs

```diff
@@ -55,20 +54,17 @@ public override T Get()
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private T GetViaScan()
         {
-            ObjectWrapper[] items = _items;
-            T item = null;
-
+            var items = _items;
             for (var i = 0; i < items.Length; i++)
             {
-                item = items[i];
-
+                var item = items[i].Element;
                 if (item != null && Interlocked.CompareExchange(ref items[i].Element, null, item) == item)
                 {
-                    break;
+                    return item;
                 }
             }
 
-            return item ?? Create();
+            return Create();
```

Comment:
This was the bug. If the `CompareExchange` fails due to contention when `i == items.Length -1` then you'll fall out of the loop and `item` will still be assigned a non-null value.

Reply:
Add this a comment?

Reply:
Was this written with a single return statement to promote inlining? Should we attempt to preserve that?

Reply:
The JIT doesn't see things like return `return item ?? Create()` - it sees things basic blocks and code sizes 😆

Reply:
> Add a comment explaining why it is done this way and to avoid previous way

I don't get it, I changed the code to be as idiomatic as possible. The bug from before was cause by writing things in a non-obvious 'single return statement fashion'.

Reply:
ok!

Reply:
> Was this written with a single return statement to promote inlining?

It was written that way to keep the loop as tight as possible (in asm). The JIT didn't produce ideal code in such cases. I.e. the function-epilog was placed "by the `return`", so the loop -- in the asm-sense -- got larger then necessary. This is also the reason why in _corefx_ there are some `goto`s, to place the epilogs more or less manually at the end of the function.
But there were same changes in this JIT-area, so that `return`s are placed outside of loops and so this problem should be solved (but I'll recheck this here).

In the hunt for "perfect" asm, the bug got introduced. Sorry for that, and thanks for finding and fixing it.

Reply:
No worries, I also signed off on the change, bugs happen.